examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,129 @@
+// -*- mode: Bluespec; -*-
+
+/**
+ * A specification ICS20 fungible token transfer protocol.
+ *
+ * For now, this only models the denomination trace changes between transfers.
+ *
+ * Gabriela Moreira, Informal Systems, 2023
+ */
+module ics20 {
+  type Uint = int
+
+  type State = {
+    trace: List[str],
+    denom: str,
+  }
+
+  type FungibleTokenPacketData = {
+    trace: List[str],
+    denom: str,
+    amount: Uint,
+    sender: str,
+    receiver: str,
+  }
+
+  type Packet = {
+    data: FungibleTokenPacketData,
+    sourcePort: str,
+    sourceChannel: str,
+    destPort: str,
+    destChannel: str,
+  }
+
+  pure def onRecvPacket(packet: Packet): State =
+    pure val trace = packet.data.trace
+    pure val sourcePrefix = [packet.sourcePort, packet.sourceChannel]
+    if (trace.length() >= 2 and trace.slice(0, 2) == sourcePrefix) {
```

Comment:
Would it make sense to introduce a `pure def` (a predicate) that tells us whether a pair `(port, channel)` are the source for an IBC denom?

Reply:
I think it would be cool to have `isSink`/`isSource` predicates.
Personally, I find them confusing, but they are the basic building blocks in ICS-020.
People could use these to debug their understanding in REPL.

Reply:
Yes! When you suggested isolating this into its own spec, I though we should define something like that so the actual ICS20 spec can call those and decide whether to mint/burn tokens. I'll try to think about a nice interface.